hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientInvocationServiceSupport.java

```diff
@@ -176,6 +178,9 @@ public void cleanResources(ConstructorFunction<Object, Throwable> responseCtor,
                 invocation.notifyException(responseCtor.createNew(null));
             }
         }
+        //Try cleaning again later to compensate the Race between `send` and `cleanResources`
```

Comment:
Perhaps in the 3.7 release, we get rid of the CHM and use an array. You want to control the number of pending invocations anyway; so for regular invocations it is best to store them in an array. Scanning invocations is then as cheap as scanning an array. 

This approach is going to be used for the improved InvocationRegistry on the server; probably the same approach can be applied on the client. 

It will also immediately solve your back pressure issues.